### DISCUSSION TITLE
Is there a way to get the last executed command?

### DISCUSSION THREAD

**Discussion by joshestein** [JetBrains/ideavim#880] [Q&A]
I am trying to find a way to see whatever the last executed command was. For example, if I do `ciw` I am trying to search backwards for the `c` command. Or if I do `fJ;;;` I am trying to get the `f` command.

I see there is `VimPlugin.HistoryGroup` but I'm not sure how to use this.

**Comment by citizenmatt** [comment]
I don't think these commands are stored as part of history. The last command is remembered internally so it can be repeated with `.`, but there isn't a history of these commands. The history group is used to handle search and Ex command history, so you can press up and down in the `/` and `:` text fields to see history.

**Comment by joshestein** [comment]
I see, thank you for the info! I'm trying to fix this incompatibility between Quickscope and Sneak plugins: https://github.com/Mishkun/ideavim-sneak/issues/27

The problem is that Sneak over-writes `f/F/t/T` mappings to reset an internal flag, `Util.lastSDirection` (setting it to null). When Quickscope is loaded after Sneak it again over-writes `f/F/t/T` mappings and so the flag will never be reset. This means whenever `;` or `,` is pressed the `SneakRepeatHandler` sees `lastSDirection` as not null (https://github.com/JetBrains/ideavim/blob/1616aff5a1711c75141c819bf6036f69fc5e5ee6/src/main/java/com/maddyhome/idea/vim/extension/sneak/IdeaVimSneakExtension.kt#L106) and Sneak moves the cursor to the last `s/S` match.

I had two thoughts about how to fix this:

1. Somehow reset the flag from inside Quickscope. I'm not sure this is possible.
2. Within Sneak, rather than checking `lastSDirection` inspect the history and see if the last command was `f/F/t/T` or `s/S`. This would allow compatibility between the two plugins.

I'm not sure if you have any other ideas or solutions? I'd appreciate any help in getting these two working together 😄

**Reply by citizenmatt** [reply]
I haven't looked at this in any detail, but do you know if the original Vim plugins have the same problem? If not, might be helpful to see how they avoid or handle the scenario - maybe that would also work in IdeaVim.

**Reply by joshestein** [reply]
That is good thinking!

I realised the [same problem already exists within Vim](https://github.com/justinmk/vim-sneak/issues/114). So I think the solution would be creating a `reset` function within Sneak, which could then be called by Quickscope on `f/F/t/T`. Is there an API to call another extension's command within IdeaVim?

**Reply by citizenmatt** [reply]
Calling another command isn't exactly straight forward, but essentially, you need to call `injector.commandGroup.getAliasCommand(name, 1)` to get a command (with count of `1`), which will either be an `Ex` command alias entered by the user, which needs to be parsed and executed as Vimscript, or a `Call` command alias, which are implemented by extensions and are simple calls to JVM code. I can't promise this API won't get refactored, as it's not very friendly.

See here:
https://github.com/JetBrains/ideavim/blob/71bcee2a0782ca3d346487f1627dbf0d1c42a733/vim-engine/src/main/kotlin/com/maddyhome/idea/vim/vimscript/model/commands/UnknownCommand.kt#L41-L61

There is an API to export a JVM function as a Vimscript function, so it would be possible to then write a mapping that would `call` a named function in your `~/.ideavimrc`. I don't know if this would be more useful for this scenario.

https://github.com/JetBrains/ideavim/blob/a81d54eb908fbb9e341122809e1e3a2bc06b3266/src/main/java/com/maddyhome/idea/vim/extension/VimExtensionFacade.kt#L219-L254

**Reply by citizenmatt** [reply]
Interestingly, I had a wander through some of the linked issues about this problem in the Vim plugins, and [one of the comments](https://github.com/unblevable/quick-scope/issues/2#issuecomment-130936366) mentions a patch to Vim that would allow scripts to know what character was last used by `f`/`F`/`t`/`T` and direction, etc. which would be pretty helpful here. Unfortunately, the link is broken, but with a bit of luck, I found [the thread on Google Groups](https://groups.google.com/g/vim_dev/c/USGJX8NBM9c/m/1p3JR96-fy4J). It suggests using the `getcharsearch` and `setcharsearch` functions, which allow a script to get the last used character, direction, etc. of a character search, and to set it too. This seems like a nice solution, if both plugins used it. I.e. sneak would read from this value to get the last char/direction, and quick scope could write to it to reset it. Ironically, neither of the Vim plugins use it.

Also, IdeaVim doesn't implement these functions 😁. But they could be added...
